### PR TITLE
[Fix #591] Enable Semantic support for C/C++; add Srefator

### LINKED ISSUE #591: Integrate Semantic Refactor for C/C++ layer
[The package](https://github.com/tuhdo/semantic-refactor) adds refactoring support for C/C++.

### PR DESCRIPTION
**Pull request by tuhdo**
With Semantic mode enabled, it enables IDE features such as:

- Enable helm-semantic-or-imenu to jump with Semantic instead of Imenu.

- Show function signature of current function point is inside at the
top (with global-semantic-stickyfunc-mode)

- Show function signature of current function at point at the
bottom (with global-semantic-idle-summary-mode).

- Add Srefactor for refactoring using Semantic parser framework.

### CODE REVIEW

**Comment by tuhdo** [comment]
I tried to set leader key for it but didn't work:

```elisp
(defun c-c++/init-cc-mode ()
  (use-package cc-mode
    :defer t
    :init
    (progn
      (semantic-mode 1)
      (add-to-list 'semantic-default-submodes 'global-semantic-stickyfunc-mode)
      (add-to-list 'semantic-default-submodes 'global-semantic-idle-summary-mode)
      (use-package srefactor
        :defer t
        :config
        (progn
          (spacemacs/declare-prefix "mr" "c-c++-refactor")
          (evil-leader/set-key-for-mode 'c++-mode
            "mrr" 'srefactor-refactor-at-point)))
      (add-hook 'c-mode-hook '(lambda () (c-toggle-auto-state t)))
      (add-hook 'c++-mode-hook '(lambda () (c-toggle-auto-state t))))))
```

**Comment by syl20bnr** [comment]
Great PR! :+1: 

Are you sure that `'c++-mode` is the right major-mode of the buffer ?
You can remove `spacemacs/declare-prefix`, it does not support per major-mode prefixes for now.

With `use-package`, the `:init` property is evaluated when `use-package` is evaluated. But `:config` is evaluated only when the package is effectively loaded. So **very important**, put your hook setup and bindings setup in a `:init` property, not `:config`.

You can join this code snippet to the PR so we can comment directly on it incrementally.

**Comment by syl20bnr** [comment]
Actually the bindings can be put in `:config` as long as the hooks are setup in `:init` and I see that your hooks are outside the `use-package` sexp so you should be OK.
Maybe the name of the major-mode then.

**Comment by syl20bnr** [comment]
You should not nest `use-package` sexps. Create a function `c-c++/init-srefactor` instead with the hook on cc-mode to enable it in `:init` and your bindings either in `:init` or `:config` (`:config` may be preferred).

**Comment by tuhdo** [comment]
I've updated the PR, but just for reviewing. Changing to `:init` still does not work. I'm sure that the modes are `c-mode` and `c++-mode`. You can open a C/C++ file, press <kbd>M-:</kbd> and type `major-mode`.

**Comment by tuhdo** [comment]
Another problem is that I just want <kbd>mr</kbd> to active the command `srefactor-refactor-at-point`, but it seems like <kbd>mr</kbd> is a group key binding and not possible.

File: contrib/lang/c-c++/packages.el [hunk 1]
```diff
@@ -22,19 +23,34 @@ which require an initialization must be listed explicitly in the list.")
 (defun c-c++/init-cc-mode ()
   (use-package cc-mode
     :defer t
-    :config
+    :init
```
Comment (syl20bnr) [inline contrib/lang/c-c++/packages.el]: Sorry my mistake, this should go under `:config`, we don't want to enable semantic when emacs starts.

File: contrib/lang/c-c++/packages.el [hunk 2]
```diff
@@ -22,19 +23,34 @@ which require an initialization must be listed explicitly in the list.")
...
+      (semantic-mode 1)
+
+      (add-to-list 'semantic-default-submodes 'global-semantic-stickyfunc-mode)
+      (add-to-list 'semantic-default-submodes 'global-semantic-idle-summary-mode)
+
       (add-hook 'c-mode-hook '(lambda () (c-toggle-auto-state t)))
       (add-hook 'c++-mode-hook '(lambda () (c-toggle-auto-state t))))))
 
+(defun c-c++/init-srefactor ()
+  (when (not (version< emacs-version "24.4"))
```
Comment (syl20bnr) [inline contrib/lang/c-c++/packages.el]: you can use `unless` instead of `when not`

File: contrib/lang/c-c++/packages.el [hunk 3]
```diff
@@ -22,19 +23,34 @@ which require an initialization must be listed explicitly in the list.")
...
+    (use-package srefactor
+      :init
```
Comment (syl20bnr) [inline contrib/lang/c-c++/packages.el]: in `use-package` there is a keyword `:if` to conditionally enable a package, you can move the tests of emacs version there:

```elisp
:if (version< 24.3 emacs-version)
```

**Comment by syl20bnr** [comment]
<kbd>mr</kbd> does not work because you wrote <kbd>mrr</kbd> ;-)

**Comment by tuhdo** [comment]
It should work properly now. `compile` module needs to be loaded otherwise this error happens:

```text
Debugger entered--Lisp error: (void-variable compilation-mode-map)
  (define-key compilation-mode-map (kbd dotspacemacs-leader-key) nil)
  (progn (define-key compilation-mode-map (kbd dotspacemacs-leader-key) nil) (define-key compilation-mode-map (kbd "h") nil))
  (lambda nil (progn (define-key compilation-mode-map (kbd dotspacemacs-leader-key) nil) (define-key compilation-mode-map (kbd "h") nil)))()
  funcall((lambda nil (progn (define-key compilation-mode-map (kbd dotspacemacs-leader-key) nil) (define-key compilation-mode-map (kbd "h") nil))))
  mapc(funcall ((lambda nil (progn (define-key compilation-mode-map (kbd dotspacemacs-leader-key) nil) (define-key compilation-mode-map (kbd "h") nil)))))
  do-after-load-evaluation("/usr/local/share/emacs/24.4.90/lisp/cedet/srecode/compile.elc")
  require(srecode/compile)
  byte-code("\300\301!\210\300\302!\210\300\303!\210\300\304!\210\300\305!\210\306\307\310\311\312\313\314\315&\207" [require srecode/compile srecode/find srecode/dictionary srecode/args srecode/filters custom-declare-variable srecode-insert-ask-variable-method (quote ask) "Determine how to ask for a dictionary value when inserting a template.\nOnly the ASK style inserter will query the user for a value.\nDictionary value references that ask begin with the ? character.\nPossible values are:\n  'ask   - Prompt in the minibuffer as the value is inserted.\n  'field - Use the dictionary macro name as the inserted value,\n           and place a field there.  Matched fields change together.\n\nNOTE: The field feature does not yet work with XEmacs." :group srecode :type (choice (const :tag "Ask" ask) (const :tag "Field" field))] 8)
  require(srecode/insert)
  byte-code("\300\301!\210\300\302!\210\300\303!\210\300\304!\210\300\305!\210\300\306!\210\307\310\311\312#\210\313\314\315\314\316\"\"\210\317\314\320\321\322$\207" [require srecode/insert srecode/dictionary semantic/find semantic/format semantic/senator ring byte-code "\300\301\302\303\304$\207" [eieio-defclass srecode-semantic-tag (srecode-dictionary-compound-value) ((prime :initarg :prime :type semantic-tag :documentation "This is the primary insertion tag.")) ("Wrap up a collection of semantic tag information.\nThis class will be used to derive dictionary values.")] 5 eieio--defalias srecode-compound-toString eieio--defgeneric-init-form "Convert the compound dictionary value CP to a string.\nIf FUNCTION is non-nil, then FUNCTION is somehow applied to an\naspect of the compound value." eieio--defmethod nil srecode-semantic-tag #[(cp function dictionary) "\204
```

**Comment by mijoharas** [comment]
Just a request, could we put semantic-mode behind an option (fine if on is the default), I'm all for it in general, but it segfaults on my work machine when I look at c++ files (I think it's a windows/cygwin/visual-studio/god-knows-what compatibility thing, I tracked it down to semantic trying to parse some weird header file, but couldn't figure out more than that).

:+1: for the PR though, looks like it will be handy, I can add the configuration option myself if you reckon it sounds like a good idea?

**Comment by tuhdo** [comment]
@mijoharas Weird. Are you sure it's Semantic's fault? i.e. you should try with stock `emacs -q` with only `semantic-mode` enabled and visit the file again, so we can confirm it's Semantic, not some 3rd party package from Spacemacs. Also, could you tell me what's your Emacs version?

> :+1: for the PR though, looks like it will be handy, I can add the configuration option myself if you reckon it sounds like a good idea?

Sure, you can add it.

**Comment by mijoharas** [comment]
TL;DR: I think I was hitting an out-of-memory error with my cygwin emacs. by breaking the parsing work done by semantic into manageable chunks by hitting `C-g` periodically I stopped it from crashing my machine. A lot of this could be wild suppositions on my part, but it makes sense to me (and things work now)

Hmmm... So I dived back into debugging what was the problem here, my best guess is some kind of cygwin out of memory error. So I couldn't repro with just `semantic-mode` as it only parsed the file I was in, however, if I opened a file that had an include that would include lots of other header files, e.g.:

```
#include <Windows.h>
```

If I entered a file with just that (with spacemacs loaded with `company-mode` and `semantic-mode` active) and typed a letter or two, it would kick off company-mode autocomplete. This uses a semantic-mode backend which parses your include files and any include files they might have. On my machine that runs for a bit and then kills emacs. This led me to a (hacky) solution to my problem.

I thought it was hitting some weird file in the tree of headers and while parsing it semantic-mode dies. I don't think this is the case, I just think if semantic-mode parses too many files at once (on my machine, cygwin emacs 32-bit cygwin) it kills itself. On trying to locate a file that kills it I would press `C-g` periodically every 5-10 files it parsed to force the list to be backed up to the messages buffer (so I could copy it and know "how far" it had gotten before dying). This stopped it from parsing too many files at once so that we got past the OOM error (or whatever it was). Since semantic-mode then writes it to the semantic-db once it parses it I can then safely open a file containing the `#include <windows.h>` line and it won't die since it just refers to the `semanticdb`

**Comment by tuhdo** [comment]
I never use the Emacs that comes with Cygwin and I think it's terrible. You should try [this 64-bit version](http://emacsbinw64.sourceforge.net/). Just download and unpack and use. No hassle. CEDET that comes with Emacs 24.3 is old and it will recursively try to parse not only current file but files that current file includes. CEDET in Emacs 24.4 does not have this problem, and Semantic Refactor only works well with CEDET in Emacs 24.4.

**Comment by mijoharas** [comment]
I'm on `GNU Emacs 24.4.1 (i686-pc-cygwin)`. For that 64-bit version you have, does it work with cygwin? the main reason I used the cygwin-emacs was in order to get the cygwin shell to be the default shell (so that all my `SPC !` and `SPC a s i` commands work as I would like them to). I struggled setting that up before when I tried to use a different emacs, how do you do that with the binary you suggested?

**Comment by tuhdo** [comment]
For Windows, I use `eshell`. Never tried Cygwin shell with it but at least, you still get all the binary from Cygwin if you add its path to Windows's `PATH` environment variable.

**Comment by mijoharas** [comment]
Thanks for the advice, I'll give it a go.

**Comment by syl20bnr** [comment]
Excellent ! :+1: 

Thank you for the great addition to the C++ layer.
Cherry-picked in `develop` branch, you can safely delete your branch.
